src/main/java/org/opensearch/neuralsearch/plugin/query/NeuralQueryBuilder.java

```diff
@@ -172,10 +193,32 @@ private static void parseQueryParams(XContentParser parser, NeuralQueryBuilder n
         }
     }
 
+    @Override
+    protected QueryBuilder doRewrite(QueryRewriteContext queryRewriteContext) {
+        if (vectorSupplier() != null && vectorSupplier.get() != null) {
+            return new KNNQueryBuilder(fieldName(), vectorSupplier.get(), k());
+        }
+
+        if (vectorSupplier() == null) {
+            SetOnce<float[]> vectorSetOnce = new SetOnce<>();
+            queryRewriteContext.registerAsyncAction(
+                ((client, actionListener) -> ML_CLIENT.inferenceSentence(modelId(), queryText(), ActionListener.wrap(floatList -> {
+                    vectorSetOnce.set(vectorAsListToArray(floatList));
+                    actionListener.onResponse(null);
```

Comment:
why setting null on the OnReponse?

Reply:
This was how it was done for Geo: https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/index/query/AbstractGeometryQueryBuilder.java#L519. 

The [rewrite context](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/index/query/QueryRewriteContext.java#L113) specifies listener response type as wildcard. That being said, we dont control the listener that is passed in, so all we can do is return null.

Reply:
I think the reason of making onResponse(null) is because we have already set the response to vector supplier. 

Can we add some more details around how this rewrite query function will work with first if condition like vectorSupplier != null and all.

Because what I am getting from this whole function is the re-writeQuery function can be called atleast 2 times. where one time vectorSupplier will be null and next time it will not be null. is that understanding correct?

Reply:
> I think the reason of making onResponse(null) is because we have already set the response to vector supplier.

We are not setting the response, but instead setting the value of the supplier.

> Can we add some more details around how this rewrite query function will work with first if condition like vectorSupplier != null and all.
> Because what I am getting from this whole function is the re-writeQuery function can be called atleast 2 times. where one time vectorSupplier will be null and next time it will not be null. is that understanding correct?

Yes, will add additional context in the comment. From my understanding, the query will be rewritten until the rewrite method returns the object itself. So, on first rewrite, the supplier will be null. On second rewrite, the supplier will be set, but the vector may be null, depending if the async call finished in time or not. Let me double check this though and add a comment.

Reply:
> > I think the reason of making onResponse(null) is because we have already set the response to vector supplier.
> 
> We are not setting the response, but instead setting the value of the supplier.
> 

I mean here the value only.

Reply:
> > Can we add some more details around how this rewrite query function will work with first if condition like vectorSupplier != null and all.
> > Because what I am getting from this whole function is the re-writeQuery function can be called atleast 2 times. where one time vectorSupplier will be null and next time it will not be null. is that understanding correct?
> 
> Yes, will add additional context in the comment. From my understanding, the query will be rewritten until the rewrite method returns the object itself. So, on first rewrite, the supplier will be null. On second rewrite, the supplier will be set, but the vector may be null, depending if the async call finished in time or not. Let me double check this though and add a comment.

My understanding is if we keep on returning the NeuralQueryBuilder, then rewrite will keep on happening(which we are doing). 

To stop query re-write we added `vectorSupplier() != null && vectorSupplier.get() != null) ` condition so that we can return another QueryBuilder(KNNQueryBuilder) which only happens when we have response from MLCommonsClientAccessor. 

Please confirm this understanding is correct or not.

Reply:
> On second rewrite, the supplier will be set, but the vector may be null, depending if the async call finished in time or not. Let me double check this though and add a comment.
I think when second rewrite happens, the supplier vector shouldn't be null, because, the second call will only happen when the async call complete because the for loop is returned [here](https://github.com/opensearch-project/OpenSearch/blob/e44158d4d10d4f8905895ffa50bf9398b8550667/server/src/main/java/org/opensearch/index/query/Rewriteable.java#L135). Once the recursive call is complete and a KNNQueryBuilder is returned, the [passed-in listener](https://github.com/opensearch-project/OpenSearch/blob/e44158d4d10d4f8905895ffa50bf9398b8550667/server/src/main/java/org/opensearch/action/search/TransportSearchAction.java#L388) will be executed and continue the query flow.

Reply:
> My understanding is if we keep on returning the NeuralQueryBuilder, then rewrite will keep on happening(which we are doing).
>
> To stop query re-write we added vectorSupplier() != null && vectorSupplier.get() != null) condition so that we can return another QueryBuilder(KNNQueryBuilder) which only happens when we have response from MLCommonsClientAccessor.
>
> Please confirm this understanding is correct or not.

Yes this is correct.

> I think when second rewrite happens, the supplier vector shouldn't be null, because, the second call will only happen when the async call complete because the for loop is returned here. Once the recursive call is complete and a KNNQueryBuilder is returned, the passed-in listener will be executed and continue the query flow.

Right, thats correct, but I return a copy at the end instead of "this" to prevent a case where rewrite stops early.